### PR TITLE
[nat64] configure nat64 prefix in Tayga

### PR DESCRIPTION
**Pull request by xuyirio**
This commits configures nat64 prefix /etc/tayga.conf when starting border agent. Tayga is restarted after configuration.

The _nat64 script is also updated to be compatible with prefix configuration by otbr-agent.

OTBR_TAYGA_NAT64 & OTBR_ENABLE_TAYGA_NAT64 are defined to guard the enablement of configuration.

Tests will be updated in a following PR.

### CODE REVIEW

File: src/nat/tayga_helper.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+{
+    otError     error = OT_ERROR_NONE;
+    otIp6Prefix nat64Prefix;
+
+    error = otBorderRoutingGetNat64Prefix(aInstance, &nat64Prefix);
+    if (error == OT_ERROR_NONE)
+    {
+        Ip6Prefix prefix;
+
+        SystemUtils::ExecuteCommand("sudo sed -i '/^prefix /d' %s", kTaygaConf);
```
Comment (wgtdkp) [inline src/nat/tayga_helper.cpp:50]:
```suggestion
        SystemUtils::ExecuteCommand("sed -i '/^prefix /d' %s", kTaygaConf);
```

And please log if `ExecuteCommand()` fails.
The same for following lines :)

Comment (xuyirio) [reply src/nat/tayga_helper.cpp]: Updated. Thanks!

File: src/nat/tayga_helper.cpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+{
+    otbrError   error = OTBR_ERROR_NONE;
+    otIp6Prefix nat64Prefix;
+    Ip6Prefix   prefix;
+
+    VerifyOrExit(otBorderRoutingGetNat64Prefix(aInstance, &nat64Prefix) == OT_ERROR_NONE,
+                 error = OTBR_ERROR_OPENTHREAD);
+
+    prefix.Set(nat64Prefix);
+    VerifyOrExit(SystemUtils::ExecuteCommand("sed -i '/^prefix /d' %s", kTaygaConf) == 0, error = OTBR_ERROR_ERRNO);
```
Comment (bukepo) [inline src/nat/tayga_helper.cpp:52]: This depends on commands may not be present on the system.

File: src/nat/tayga_helper.cpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    VerifyOrExit(
+        SystemUtils::ExecuteCommand("sed -i '/^##prefix /a prefix '%s %s", prefix.ToString().c_str(), kTaygaConf) == 0,
+        error = OTBR_ERROR_ERRNO);
+
+    VerifyOrExit(SystemUtils::ExecuteCommand("systemctl restart tayga") == 0, error = OTBR_ERROR_ERRNO);
```
Comment (bukepo) [inline src/nat/tayga_helper.cpp:57]: The system may not use systemd to manage services.

File: src/nat/tayga_helper.cpp (hunk 2 above)
Comment (bukepo) [reply src/nat/tayga_helper.cpp:52]: Perhaps you can use [libconfig](http://www.hyperrealm.com/libconfig/libconfig_manual.html) instead.

Comment (xuyirio) [reply src/nat/tayga_helper.cpp:52]: Thanks for point out. This nat64 prefix configuration will be enabled when border routing and nat64 are both supported, for which we expect to be ubuntu and raspbian as of now. And the sed command is expected to work just as in the _nat64 script.
libconfig usually works on structured conf files

Comment (bukepo) [reply src/nat/tayga_helper.cpp:52]: The scripts are supposed to run on ubuntu and rasbian, but the c++ binaries are not limited to these platforms.

Comment (simonlingoogle) [reply src/nat/tayga_helper.cpp:52]: > Perhaps you can use [libconfig](http://www.hyperrealm.com/libconfig/libconfig_manual.html) instead.

Using `libconfig` seems too much dependency for just changing one line in a file?

Maybe we can generate the whole `tayga.conf` file from scratch in C++ code without using any system command or library?

File: src/nat/tayga_helper.cpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
+/*
+ *    Copyright (c) 2017, The OpenThread Authors.
```
[the remark this answers is not in the document]
Comment (xuyirio) [reply src/nat/tayga_helper.cpp]: Updated

File: src/nat/tayga_helper.hpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
+/*
+ *    Copyright (c) 2017, The OpenThread Authors.
```
[the remark this answers is not in the document]
Comment (xuyirio) [reply src/nat/tayga_helper.hpp]: Updated

[the remark this answers is not in the document]
Comment (xuyirio) [reply src/nat/tayga_helper.hpp]: Was updated.

File: src/nat/tayga_helper.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (xuyirio) [reply src/nat/tayga_helper.cpp]: Updated to also try "service tayga restart" (e.g. in DOCKER).

File: src/nat/tayga_helper.hpp [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * @file
+ *   The file implements helper functions for TAYGA.
+ */
+
+#ifndef OTBR_AGENT_TAYGA_HELPER_HPP_
+#define OTBR_AGENT_TAYGA_HELPER_HPP_
+
+#include <openthread/instance.h>
```
[the remark this answers is not in the document]
Comment (xuyirio) [reply src/nat/tayga_helper.hpp:37]: Looks that we are doing this. E.g. in src/utils/thread_helper.hpp

File: src/nat64/tayga_helper.cpp [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+ *    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <fstream>
```
Comment (simonlingoogle) [inline src/nat64/tayga_helper.cpp:29]: `#include <fstream>` should be inside `#if OTBR_ENABLE_NAT64` ?

File: src/nat64/tayga_helper.cpp [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+{
+    bool          is_updated = false;
+    std::ifstream file(kTaygaConf);
+    std::ofstream tmp_file(kTaygaConfTmp);
+    std::string   line;
+
+    if (!file || !tmp_file)
+    {
+        otbrLogErr("Failed to open Tayga configure files");
+        return is_updated;
```
Comment (simonlingoogle) [inline src/nat64/tayga_helper.cpp:55]: we only allow one return for each function/method.

File: src/nat64/tayga_helper.cpp [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    }
```
Comment (simonlingoogle) [inline src/nat64/tayga_helper.cpp:56]:
```suggestion
    VerifyOrExit(file && tmp_file, otbrLogErr("Failed to open Tayga configure files"));
```

File: src/nat64/tayga_helper.cpp [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+    while (getline(file, line))
```
Comment (simonlingoogle) [inline src/nat64/tayga_helper.cpp:58]: handle retval -1?

File: src/nat64/tayga_helper.cpp [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        if (line.find(kPrefixLineStart, 0) == 0)
+        {
+            line       = kPrefixLineStart + prefix;
+            is_updated = true;
+        }
+        tmp_file << line << std::endl;
+    }
+
+    file.close();
+    tmp_file.close();
```
Comment (simonlingoogle) [inline src/nat64/tayga_helper.cpp:69]: need to make sure both files are closed whenever this function returns.

File: src/nat64/tayga_helper.cpp [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+    if (is_updated)
+    {
+        remove(kTaygaConf);
```
Comment (simonlingoogle) [inline src/nat64/tayga_helper.cpp:73]: handle failure?

File: src/nat64/tayga_helper.cpp [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        rename(kTaygaConfTmp, kTaygaConf);
```
Comment (simonlingoogle) [inline src/nat64/tayga_helper.cpp:74]: handle failure?

Comment (simonlingoogle) [inline src/nat64/tayga_helper.cpp:74]: handle failure?

Comment (simonlingoogle) [inline src/nat64/tayga_helper.cpp:74]: rename will automatically replace `kTaygaConf` if it exists. So, no need to `remove` it first?

File: src/nat64/tayga_helper.cpp (hunk 7 above)
Comment (xuyirio) [reply src/nat64/tayga_helper.cpp:30]: Updated.

File: src/nat64/tayga_helper.cpp (hunk 8 above)
Comment (xuyirio) [reply src/nat64/tayga_helper.cpp]: Updated.

File: src/nat64/tayga_helper.cpp (hunk 9 above)
Comment (xuyirio) [reply src/nat64/tayga_helper.cpp]: Updated.

File: src/nat64/tayga_helper.cpp (hunk 10 above)
Comment (xuyirio) [reply src/nat64/tayga_helper.cpp:57]: Added checking for unexpected stream state.

File: src/nat64/tayga_helper.cpp (hunk 11 above)
Comment (xuyirio) [reply src/nat64/tayga_helper.cpp:71]: Updated.

File: src/nat64/tayga_helper.cpp (hunk 12 above)
Comment (xuyirio) [reply src/nat64/tayga_helper.cpp]: Added.

File: src/nat64/tayga_helper.cpp (hunk 13 above)
Comment (xuyirio) [reply src/nat64/tayga_helper.cpp]: Added.

Comment (xuyirio) [reply src/nat64/tayga_helper.cpp]: rename might fail for existing file name in some system. So call remove first to make it safer.

File: src/CMakeLists.txt [hunk 14]
```diff
@@ -33,6 +33,7 @@ endif()
 add_subdirectory(common)
 add_subdirectory(ncp)
 add_subdirectory(sdp_proxy)
+add_subdirectory(nat)
```
[the remark this answers is not in the document]
Comment (xuyirio) [reply src/CMakeLists.txt]: Updated.

File: src/nat/tayga_helper.cpp (hunk 2 above)
Comment (xuyirio) [reply src/nat/tayga_helper.cpp]: Updated to copy and rewrite the file.

File: src/nat/tayga_helper.cpp [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+                 error = OTBR_ERROR_OPENTHREAD);
+
+    prefix.Set(nat64Prefix);
+    VerifyOrExit(SystemUtils::ExecuteCommand("sed -i '/^prefix /d' %s", kTaygaConf) == 0, error = OTBR_ERROR_ERRNO);
+    VerifyOrExit(
+        SystemUtils::ExecuteCommand("sed -i '/^##prefix /a prefix '%s %s", prefix.ToString().c_str(), kTaygaConf) == 0,
+        error = OTBR_ERROR_ERRNO);
+
+    VerifyOrExit(SystemUtils::ExecuteCommand("systemctl restart tayga") == 0 ||
+                     SystemUtils::ExecuteCommand("service tayga restart") == 0,
```
[the remark this answers is not in the document]
Comment (xuyirio) [reply src/nat/tayga_helper.cpp]: Thanks for the suggestion. Added an option OTBR_TAYGA_RESTART_CMD for configuring the command. Its default value is set to "systemctl restart tayga".

File: src/nat64/tayga_helper.cpp [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+{
+    bool          is_copied  = false;
+    bool          is_updated = false;
+    std::ifstream file(kTaygaConf);
+    std::ofstream tmp_file(kTaygaConfTmp);
+    std::string   line;
+
+    VerifyOrExit(file && tmp_file);
+
+    while (getline(file, line))
```
Comment (simonlingoogle) [inline src/nat64/tayga_helper.cpp:58]:
```suggestion
    while (std::getline(file, line))
```
nit: reduce confusion

File: src/nat64/tayga_helper.cpp [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+        if (line.find(kPrefixLineStart, 0) == 0)
+        {
+            line       = kPrefixLineStart + prefix;
+            is_updated = true;
+        }
+        tmp_file << line << std::endl;
+        VerifyOrExit(tmp_file.good(), is_updated = false);
+    }
+
+    VerifyOrExit(file.eof(), is_updated = false);
```
Comment (simonlingoogle) [inline src/nat64/tayga_helper.cpp:69]: may fail to remove tmp file if exit earlier?

File: src/nat64/tayga_helper.cpp [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+
+    file.close();
+    tmp_file.close();
+    is_copied = true;
+
+    if (is_updated)
+    {
+        VerifyOrExit(remove(kTaygaConf) == 0 && rename(kTaygaConfTmp, kTaygaConf) == 0, is_updated = false);
```
Comment (simonlingoogle) [inline src/nat64/tayga_helper.cpp:77]:
```suggestion
        VerifyOrExit(rename(kTaygaConfTmp, kTaygaConf) == 0, is_updated = false);
```
`remove` not necessary?

File: src/nat64/tayga_helper.cpp [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+namespace otbr {
+namespace Tayga {
+namespace {
+
+static const char kTaygaConf[]       = "/etc/tayga.conf";
+static const char kTaygaConfTmp[]    = "/etc/tayga_tmp.conf";
+static const char kTaygaRestartCmd[] = OTBR_TAYGA_RESTART_CMD;
+static const char kPrefixLineStart[] = "prefix ";
+
+bool UpdatePrefix(const std::string &prefix)
```
Comment (simonlingoogle) [inline src/nat64/tayga_helper.cpp:48]:
```suggestion
otbrError UpdatePrefix(const std::string &prefix)
```
nit: we commonly return errors

File: src/nat64/tayga_helper.cpp (hunk 17 above)
Comment (xuyirio) [reply src/nat64/tayga_helper.cpp:69]: Updated to remove tmp file on exit

File: src/nat64/tayga_helper.cpp (hunk 18 above)
Comment (xuyirio) [reply src/nat64/tayga_helper.cpp:77]: `rename` might fail for exiting files in some system. So call `remove` to make it safer.

File: src/nat64/tayga_helper.cpp (hunk 19 above)
Comment (xuyirio) [reply src/nat64/tayga_helper.cpp]: updated

File: src/nat64/tayga_helper.cpp (hunk 16 above)
Comment (xuyirio) [reply src/nat64/tayga_helper.cpp]: update. thanks!

File: src/nat64/tayga_helper.hpp [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+#include <openthread/instance.h>
+
+namespace otbr {
+namespace Tayga {
+
+/**
+ * This function configures the nat64 prefix for Tayga.
+ *
+ * @param aInstance A pointer to an OpenThread instance.
+ */
```
[the remark this answers is not in the document]
Comment (xuyirio) [reply src/nat64/tayga_helper.hpp]: updated.

**Comment by xuyirio** [comment]
@simonlingoogle @wgtdkp @bukepo @zhanglongxia Could you help take another look at this PR?
Although we are still evaluating 1) using Tayga and integrate with 1P products vs. 2) building a new stateful NAT64 service, I'd like to get the Tayga configured with the appropriate prefix first. This will help us create a prototype quickly and make testing easier. Thanks!

**Comment by simonlingoogle** [comment]
> I'd like to get the Tayga configured with the appropriate prefix first. This will help us create a prototype quickly and make testing easier. Thanks!

+1. 
I think a TAYGA based approach can still be useful for open source solutions, at least in the short term.

File: etc/cmake/options.cmake [hunk 21]
```diff
@@ -69,7 +74,7 @@ if (OTBR_SRP_ADVERTISING_PROXY)
     target_compile_definitions(otbr-config INTERFACE OTBR_ENABLE_SRP_ADVERTISING_PROXY=1)
 endif()
 
-option(OTBR_DNSSD_DISCOVERY_PROXY   "Enable DNS-SD Discovery Proxy support" OFF)
+option(OTBR_DNSSD_DISCOVERY_PROXY "Enable DNS-SD Discovery Proxy support" OFF)
```
Comment (simonlingoogle) [inline etc/cmake/options.cmake:77]: nit: better avoid unrelated change.

File: src/CMakeLists.txt [hunk 22]
```diff
@@ -32,6 +32,7 @@ if(OTBR_BORDER_AGENT)
 endif()
 add_subdirectory(common)
 add_subdirectory(ncp)
+add_subdirectory(nat64)
```
Comment (simonlingoogle) [inline src/CMakeLists.txt:35]:
```suggestion
add_subdirectory(tayga_nat64)
```

I think this directory is dedicated for tayga?

File: src/nat64/CMakeLists.txt [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+
+add_library(otbr-nat64
```
Comment (simonlingoogle) [inline src/nat64/CMakeLists.txt:29]:
```suggestion
add_library(otbr-tayga-nat64
```
nit

File: src/nat64/tayga_helper.cpp [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+ *    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if OTBR_ENABLE_TAYGA_NAT64
+#include <fstream>
```
Comment (simonlingoogle) [inline src/nat64/tayga_helper.cpp:30]:
```suggestion
#if OTBR_ENABLE_TAYGA_NAT64

#include <fstream>
```

File: third_party/openthread/CMakeLists.txt [hunk 25]
```diff
@@ -29,6 +29,7 @@
 set(OT_BORDER_AGENT ON CACHE BOOL "enable border agent" FORCE)
 set(OT_BORDER_ROUTER ON CACHE BOOL "enable border router feature" FORCE)
 set(OT_BORDER_ROUTING ${OTBR_BORDER_ROUTING} CACHE BOOL "enable border routing feature" FORCE)
+set(OT_BORDER_ROUTING_NAT64 ${OTBR_TAYGA_NAT64}&&${OTBR_BORDER_ROUTING} CACHE BOOL "enable border routing NAT64 feature" FORCE)
```
Comment (simonlingoogle) [inline third_party/openthread/CMakeLists.txt:32]: We may still set OTBR_TAYGA_NAT64=ON when OTBR_TAYGA_NAT64 is OFF, if we are using other NAT64 implementations?

File: src/nat64/tayga_helper.cpp [hunk 26]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    otbrLogInfo("NAT64 prefix in Tayga configuration file %s updated", is_updated ? "is" : "isn't");
+
+    return is_updated;
+}
+
+} // namespace
+
+void ConfigTaygaNat64Prefix(otInstance *aInstance)
```
Comment (simonlingoogle) [inline src/nat64/tayga_helper.cpp:101]:
```suggestion
void UpdateTaygaNat64Prefix(otInstance *aInstance)
```
nit

File: src/nat64/tayga_helper.cpp [hunk 27]
```diff
@@ -0,0 +1,129 @@
...
+{
+    otbrError   error      = OTBR_ERROR_NONE;
+    bool        is_updated = false;
+    otIp6Prefix nat64Prefix;
+    Ip6Prefix   prefix;
+
+    VerifyOrExit(otBorderRoutingGetNat64Prefix(aInstance, &nat64Prefix) == OT_ERROR_NONE,
```
Comment (simonlingoogle) [inline src/nat64/tayga_helper.cpp:108]: when there there is no NAT64 prefix, should we stop tayga?

File: src/nat64/tayga_helper.cpp [hunk 28]
```diff
@@ -0,0 +1,129 @@
...
+                 error = OTBR_ERROR_OPENTHREAD);
+
+    prefix.Set(nat64Prefix);
```
Comment (simonlingoogle) [inline src/nat64/tayga_helper.cpp:111]: should we avoid restarting tayga if the NAT64 prefix is not actually changed?

File: src/nat64/tayga_helper.cpp (hunk 18 above)
Comment (simonlingoogle) [reply src/nat64/tayga_helper.cpp:79]: > `rename` might fail for exiting files in some system. So call `remove` to make it safer.

According to [manpages](https://linux.die.net/man/3/rename), the new file is always removed if exists. Do we have any clue why it may fail for existing files in some system?

Comment (xuyirio) [reply src/nat64/tayga_helper.cpp:79]: Based on the description of [rename()](https://www.cplusplus.com/reference/cstdio/rename/), `If newname names an existing file, the function may either fail or override the existing file, depending on the specific system and library implementation`. So I think first removing the existing file is safer?

File: src/nat64/tayga_helper.cpp (hunk 28 above)
Comment (xuyirio) [reply src/nat64/tayga_helper.cpp:111]: Yes. We check `is_updated` before restarting tayga.

File: third_party/openthread/CMakeLists.txt (hunk 25 above)
Comment (xuyirio) [reply third_party/openthread/CMakeLists.txt:32]: I don't quite understand. Do you mean to set `OT_BORDER_ROUTING_NAT64=ON` if we are using any other NAT64 implementation? When we have other implementations later, we can update the condition here to reflect the change.

File: src/CMakeLists.txt (hunk 22 above)
Comment (xuyirio) [reply src/CMakeLists.txt:35]: Tayga is one kind of implementation for NAT64. Other implementation can also be added into the same directory. If size of the directory becomes a concern later, we can refactor the structure accordingly.

File: src/nat64/tayga_helper.cpp (hunk 26 above)
Comment (xuyirio) [reply src/nat64/tayga_helper.cpp:101]: Since we might restart tayga, I think `config` is closer description?

File: src/nat64/tayga_helper.cpp (hunk 27 above)
Comment (xuyirio) [reply src/nat64/tayga_helper.cpp:110]: Updated to stop tayga. 

When `OT_BORDER_ROUTING_NAT64` is ON, a nat64 prefix usually exists when `ConfigTaygaNat64Prefix()` is called (routing manager is initialized). So it should be a rare case.

File: src/nat64/tayga_helper.cpp (hunk 18 above)
Comment (simonlingoogle) [reply src/nat64/tayga_helper.cpp:81]: Thanks for the reference. LGTM.

File: src/CMakeLists.txt (hunk 22 above)
Comment (simonlingoogle) [reply src/CMakeLists.txt:35]: SGTM.
